### ISSUE TITLE
Inherit not defined celebration definition metadata from `general.ts`

### ISSUE THREAD

**Issue by tukusejssirs** [pejulian/romcal#206]
Related #48

---

I suggest to inherit celebration definition metadata from `general.ts` for celebrations redefined in particular calendars.

For example:

1. A celebration is defined in `general.ts` and it is an optional memorial celebrated on 24 June.
2. In a particular calendar, the very same celebration is also celebrated, but the rank is raised to memorial, therefore we should need only to change the celebration rank and should not need to redefine the date nor anything else that is not changed in comparison with the `general.ts` calendar.

Also note that this should be the case not only for the `general.ts` calendar, but for any calendar when the calendar inheritance would be implemented (see #48).

**Comment by emagnier** [comment]
This is directly related to https://github.com/romcal/romcal/issues/48.

**Comment by tukusejssirs** [comment]
True. But this is a different suggestion. Still, these two issues should be fixed side by side. :smiley:

**Comment by emagnier** [comment]
For example, we have a celebration defined in the GRC.

```json5
// general.ts
{
  key: 'saintGregoryTheGreatPopeAndDoctor',
  rank: Ranks.MEMORIAL,
  date: dayjs.utc(`${year}-9-3`),
  liturgicalColors: LiturgicalColors.WHITE,
  metadata: {
    titles: [Titles.DOCTOR_OF_THE_CHURCH],
  },
},
```

In England, only the `rank` is different. We simply extend its definition from the `general` calendar with the `inheritsFrom` property, and add/update only what is different in this particular calendar:

```json5
// england.ts
{
  inheritsFrom: 'saintGregoryTheGreatPopeAndDoctor',
  rank: Ranks.FEAST,
},
```

Regarding the `inheritsFrom` keyword, I also have other variants or synonyms in mind, I'm not able to choose one:
- `inheritsFrom`
- `inheritFrom`
- `extends`

Basically, it means that this definition in this particular calendar `extends` / `inheritsFrom` the previous one defined in `general`. The `extends` is a keyword often uses in language programming to extend classes (including Java and TypeScript). Inherit(s) from is also often used in documentation to describe this behavior.

Note: the future calendar inheritance feature should use the same keyword for consistency (when a calendar extends / inherits from another calendar).

I might have a small preference for the `inheritsFrom` keyword, perhaps more meaningful for everyone (not only for developers). But I would like feedbacks.

**Comment by tukusejssirs** [comment]
IMO:
- `extend` word usually means to _add_ something _more_ to something else (you might to see the [explanation of all meanings of the word](https://www.lexico.com/en/definition/extend)), however, we usually don’t _add_ anything to the already defined celebration, only _replace_ some of the metadata;
- `inheritFrom` is much better IMO, it means that we _get/receive_ something from someone/somewhere, which which [the something] we could do anything we want (like derive something from it, spend the money, sell it), because it is _ours_ (you might want to see the [definition of the meanings of the word](https://www.lexico.com/en/definition/inherit));
- in `inheritFrom` vs `inheritsFrom` I prefer `inheritFrom`, because IMI it should in imperative mode (even some Git guides suggest to use [imperatives in the commit messages](https://chris.beams.io/posts/git-commit/)).

**Comment by emagnier** [comment]
FYI this has been implemented and documented here: https://github.com/romcal/romcal/blob/_F-TypeScript/docs/calendar-definitions.md#extending-a-liturgical-day

It use the `extend` keyword, but in a different way from what I suggested above. The `extend` flag is now just a boolean (`false` by default), and the `key` property is always required on each definitions. I think that way it will be less confusing.
When the extend flag is `true`, romcal will look in the general/celebration calendar, and extend the already defined celebration with the new properties defined in the particular calendar. Please have a look to the documentation for more details.

There is an example in [england.ts](https://github.com/romcal/romcal/blob/78a462dd93d7f5e0ca92b8650d0dff7961b25d5b/src/calendars/england.ts#L226-L230) on `gregory_i_the_great_pope` where the rank is raised to `feast`, but the other properties remain the same, so no need to define them again.

**Comment by tukusejssirs** [comment]
Okay, I like the implementation, however, don’t you think it should be `true` by default? Because in 99 % of celebration that are celebrated in `general` and have some celebration definition metadata changed in a particular calendar, there is only one or two changes.

Actually, I think we don’t need the `extend` keyword at all: the celebration should extended by default. When one needs to override all the metadata, they would change all the keys (keywords) anyway. :smiley:

**Comment by emagnier** [comment]
> don’t you think it should be true by default?

I took a conservative approach, but you are right, this could be `true` by default.

> Actually, I think we don’t need the extend keyword at all: the celebration should extended by default. When one needs to override all the metadata, they would change all the keys (keywords) anyway.

I agree. 
However, by removing the `extend: true` property, we must add a test to ensure that all liturgical day definitions have at least a `date` defined. Up to now this was done automatically thanks to the TypeScript typing.

**Comment by tukusejssirs** [comment]
>  However, by removing the `extend: true` property, we must add a test to ensure that all liturgical day definitions have at least a `date` defined. Up to now this was done automatically thanks to the TypeScript typing.

IMHO that test would be nice to have even when we won’t remove `extend` keyword. Still, I think it should be removed. Do as you think it is the best.

**Comment by emagnier** [comment]
FYI, thanks to the TypeScript typing, if a liturgical day isn't well defined (missing/unknown/misused property), you will not be able to build romcal, so tests will fail.

**Comment by emagnier** [comment]
This feature is now implemented in romcal v3 (from #48).

And no need of `extend: true` or similar properties (as @tukusejssirs [suggested](https://github.com/romcal/romcal/issues/206#issuecomment-735285366) 2 comment above): all definitions will automatically extend a previously definition that have the same `key`.

For example below, in the `europe` calendar, the date of `benedict_of_nursia_abbot` isn't redefined (it remains the date defined in the GRC), but only its Precedence, Title and Name that is proper to the European region.

https://github.com/romcal/romcal/blob/955dd6c01ffd68f47717002c606539ed81f7f056/lib/particular-calendars/europe.ts#L21-L25
